### PR TITLE
Reduce stack usage of the parser.

### PR DESCRIPTION
**Pull request by zherczeg**
Inline CoverGrammar functions and move the common code into helper functions.

### CODE REVIEW

File: src/parser/esprima_cpp/esprima.cpp [hunk 1]
```diff
@@ -799,140 +805,105 @@ class Parser : public gc {
         }
     }
 
-    template <typename T>
-    PassRefPtr<Node> isolateCoverGrammar(T parseFunction)
+    void startCoverGrammar(IsolateCoverGrammarContext* grammarContext)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(grammarContext != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 2]
```diff
@@ -799,140 +805,105 @@ class Parser : public gc {
...
         this->context->isAssignmentTarget = true;
         this->context->firstCoverInitializedNameError = nullptr;
 
-        this->checkRecursiveLimit();
-        PassRefPtr<Node> result = (this->*parseFunction)();
+        checkRecursiveLimit();
+    }
+
+    void endIsolateCoverGrammar(IsolateCoverGrammarContext* grammarContext)
+    {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(grammarContext != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 3]
```diff
@@ -799,140 +805,105 @@ class Parser : public gc {
...
-        this->context->isAssignmentTarget = previousIsAssignmentTarget;
-        this->context->firstCoverInitializedNameError = previousFirstCoverInitializedNameError;
+        this->context->isBindingElement = grammarContext->previousIsBindingElement;
+        this->context->isAssignmentTarget = grammarContext->previousIsAssignmentTarget;
+        this->context->firstCoverInitializedNameError = grammarContext->previousFirstCoverInitializedNameError;
+    }
 
-        return result;
+    void endInheritCoverGrammar(IsolateCoverGrammarContext* grammarContext)
+    {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(grammarContext != nullptr);

**Comment by clover2123** [comment]
Parser & Lexer is updated not to use `RefPtr` for ScannerResult. Please rebase this patch based on the master branch.

**Comment by clover2123** [comment]
BTW, it seems that this patch uses the same stack size. Does this patch really reduce the stack usage?

File: src/parser/esprima_cpp/esprima.cpp [hunk 4]
```diff
@@ -803,140 +809,111 @@ class Parser : public gc {
...
-        PassRefPtr<Node> result = (this->*parseFunction)();
-        if (UNLIKELY(this->context->firstCoverInitializedNameError)) {
+        checkRecursiveLimit();
+    }
+
+    void endIsolateCoverGrammar(IsolateCoverGrammarContext* grammarContext)
+    {
+        ASSERT(grammarContext != nullptr);
+
+        if (this->context->firstCoverInitializedNameError) {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: please state expression explicitly like ``if (this->context->firstCoverInitializedNameError != nullptr)``

File: src/parser/esprima_cpp/esprima.cpp [hunk 5]
```diff
@@ -803,140 +809,111 @@ class Parser : public gc {
...
+    }
 
-        return result;
+    void endInheritCoverGrammar(IsolateCoverGrammarContext* grammarContext)
+    {
+        ASSERT(grammarContext != nullptr);
+
+        this->context->isBindingElement = this->context->isBindingElement && grammarContext->previousIsBindingElement;
+        this->context->isAssignmentTarget = this->context->isAssignmentTarget && grammarContext->previousIsAssignmentTarget;
+        if (grammarContext->previousFirstCoverInitializedNameError) {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ditto

**Comment by zherczeg** [comment]
The idea is the core functions are `ALWAYS_INLINE` now, so no new stack frame is created for them. To reduce code size increase, the save/restore part are moved to helper functions.

File: src/parser/esprima_cpp/esprima.cpp (hunk 4 above)
Comment (zherczeg) [reply src/parser/esprima_cpp/esprima.cpp]: This member is not a pointer anymore. It has a bool() cast operator though:

```
inline operator bool() const
{
    return this->type != InvalidToken;
}
```

I could add a `->type != InvalidToken` but I am not sure it looks better.

Comment (yichoi) [reply src/parser/esprima_cpp/esprima.cpp]: Your suggention is more perpect in terms with expliciticity and readability! Please do like that. I promise you will get accustomed with new convention :)

Comment (zherczeg) [reply src/parser/esprima_cpp/esprima.cpp]: I did it. Shall I create another patch to remove all automated `bool` conversion?

Comment (yichoi) [reply src/parser/esprima_cpp/esprima.cpp]: If you can, please do however don't be hurry it's not urgent.

**Pull request merged**